libopae/plugins/xfpga/metrics/vector.c

```diff
@@ -0,0 +1,170 @@
+// Copyright(c) 2018, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+* \file vector.c
+* \brief fpga metrics vector
+*/
+
+#include "vector.h"
+#include "opae/access.h"
+#include "opae/utils.h"
+#include "opae/manage.h"
+#include "opae/enum.h"
+#include "opae/properties.h"
+#include "common_int.h"
+#include "safe_string/safe_string.h"
+
+#define FPGA_VECTOR_CAPACITY     20
+
+fpga_result fpga_vector_init(fpga_metric_vector *vector)
+{
+	fpga_result result = FPGA_OK;
+
+	if (vector == NULL)
+		return FPGA_INVALID_PARAM;
+
+	vector->capacity = FPGA_VECTOR_CAPACITY;
+	vector->total = 0;
+	vector->fpga_metric_item = malloc(sizeof(void *) * vector->capacity);
+
+	if (vector->fpga_metric_item == NULL)
+		return FPGA_NO_MEMORY;
+
+	return result;
+}
+
+fpga_result fpga_vector_free(fpga_metric_vector *vector)
+{
+	fpga_result result = FPGA_OK;
+	if (vector == NULL) {
+		FPGA_ERR("Invalid parm");
+		return FPGA_INVALID_PARAM;
+	}
+	if(vector->fpga_metric_item)
+		free(vector->fpga_metric_item);
+
+	vector->fpga_metric_item = NULL;
+
+	return result;
+}
+
+uint64_t fpga_vector_total(fpga_metric_vector *vector)
```

Comment:
Maybe the total should be an output parameter, and the function should return an fpga_result?  Isn't 0 a valid value?

Reply:
Fixed